src/raft.c

```diff
@@ -115,19 +115,185 @@ void entryAttachRaftReq(RedisRaftCtx *rr, raft_entry_t *entry, RaftReq *req)
+        }
+        return RR_ERROR;
+    }
+    unsigned int slot = (unsigned int) cmds->slot;
+
+    SlotRangeType slot_type = SLOTRANGE_TYPE_UNDEF;
+    for (size_t i = 0; i < sg->slot_ranges_num; i++) {
+        if (sg->slot_ranges[i].start_slot <= slot && slot <= sg->slot_ranges[i].end_slot) {
+            slot_type = sg->slot_ranges[i].type;
+            break;
+        }
+    }
+
+    if (slot_type == SLOTRANGE_TYPE_UNDEF) {
+        if (reply_ctx) {
+            RedisModule_ReplyWithError(reply_ctx, "ERR internal error, couldn't associate a shardgroup slot to this request");
+        }
+        return RR_ERROR;
+    }
+
+    if (!sg->local) {
+        if (reply_ctx) {
+            if (sg->next_redir >= sg->nodes_num) {
+                sg->next_redir = 0;
+            }
+
+            replyRedirect(reply_ctx, cmds->slot, &sg->nodes[0].addr);
+        }
+
+        return RR_ERROR;
+    }
+
+    /* if our keys belong to a local migrating/importing slot, all keys must exist */
+    if (slot_type == SLOTRANGE_TYPE_MIGRATING) {
+        switch (validateKeyExistence(rr, cmds)) {
+            case SomeExist:
+                if (reply_ctx) {
+                    RedisModule_ReplyWithError(reply_ctx, "TRYAGAIN");
+                }
+                return RR_ERROR;
+            case NoneExist:
+                if (reply_ctx) {
+                    replyAsk(rr, reply_ctx, cmds->slot);
+                }
+                return RR_ERROR;
+            case AllExist:
+            case NoKeys:
+                return RR_OK;
+        }
+    } else if (slot_type == SLOTRANGE_TYPE_IMPORTING) {
+        switch (validateKeyExistence(rr, cmds)) {
+            case SomeExist:
+            case NoneExist:
+                if (reply_ctx) {
+                    RedisModule_ReplyWithError(reply_ctx, "TRYAGAIN");
+                }
+                return RR_ERROR;
+            case AllExist:
+            case NoKeys:
+                return RR_OK;
+        }
+    }
+
+    return RR_OK;
+}
+
 /* Execute all commands in a specified RaftRedisCommandArray.
  *
  * If reply_ctx is non-NULL, replies are delivered to it.
  * Otherwise, no replies are delivered.
  */
-void RaftExecuteCommandArray(RedisModuleCtx *ctx,
+void RaftExecuteCommandArray(RedisRaftCtx *rr,
+                             RedisModuleCtx *ctx,
                              RedisModuleCtx *reply_ctx,
                              RaftRedisCommandArray *cmds)
 {
     int i;
 
     HandleAsking(cmds);
 
+    /* When we're in cluster mode, go through handleSharding. This will perform
+     * hash slot validation and return an error / redirection if necessary. We do this
+     * before checkLeader() to avoid multiple redirect hops.
+     */
+    if (isSharding(rr) && handleSharding(rr, reply_ctx, cmds) != RR_OK) {
+        return;
+    }
+
+    if (cmds->slot != -1 && validateRaftRedisCommandArray(rr, reply_ctx, cmds) != RR_OK) {
+        return;
+    }
```

Comment:
if we are removing handleSharding from append path, can combine the `validateRaftRedisCommandArray()` into `handleSharding()`